### PR TITLE
Extended Rearrangement

### PR DESCRIPTION
**Pull request by RLee64**
## Overview
In #1 we resolved the rearrangement example seen in #1354 for the trivial x = y + z case. However, we want to expand upon this functionality in order to support more operations. As such, this pull request focuses on growing the existing level of support for equation rearrangement, expanding functionality to other use cases. That being said, this implementation does not fully cover all libcellml supported types, but we hope to expand upon this in an additional pull request.

Here we also begin to build a set of proper test cases for us to ensure SymEngine's rearrangement functionality is behaving as we intend.

## Changes
- Added test cases for equation rearrangement (originally written by GPT, but I've adjusted files to fix errors and hopefully cover more edge cases)
- Parsing from SymEngine back to AST now supports SymEngine expressions with 3+ children
- Rearrangement support for addition and subtraction
- Rearrangement support for unary +/-
- Rearrangement support for multiplication and division
- Rearrangement support for basic trigonometric rearrangement (when the unknown variable is not inside a trig function)
- Rearrangement support for various constants (i.e. e, pi, inf)
- Rearrangement support for different number 'types' (i.e. doubles and scientific notation)
- Rearrangement support for polynomials with one **real** solution
- In order to achieve polynomial rearrangement support, the method ``isSymEngineExpressionComplex`` was introduced to recursively look for complex constituents. This is consequently used to remove complex equations from our solution set.

## Notes
- Since SymEngine occasionally throws exceptions when the system is unable to rearrange an equation (e.g. when needing to rearrange tan to atan or having to work with fractional polynomials), I'm guessing we'll probably want to catch these errors at some point to ensure that these errors don't terminate code execution just because rearrangement wasn't successful. With that in mind, I haven't done it yet since getting symengine errors is useful for discovering the library's limitations.
- Similar to the previous PR, we're once again failing some existing test cases. At a quick glance, it seems that the unordered algebraic system is running into an issue with variables now being underconstrained and the hodgkin huxley model tests now have some equations interpreted as algebraic. I'll have a deeper look into these to get a better idea of what's going on.
<div style="display:block; margin:auto;">
image
<div />

- Not all of the new test cases currently pass, but this PR is probably big enough as is, so I'm hoping to finish them off in a third branch that wraps up rearrangement (at least for the time being).

<div style="display:block; margin:auto;">
image
<div />

### CODE REVIEW

File: src/analyser.cpp
Comment (agarny) [inline src/analyser.cpp:202]: Why limit ourselves to what you have here? Can we not support the conversion of **_everything_** CellML supports and leave it to SymEngine to do whatever it can with it? In other words, I would convert everything that  is listed [here](https://github.com/RLee64/libcellml/blob/main/src/api/libcellml/analyserequationast.h#L43-L136).

File: src/analyser.cpp [hunk 1]
```diff
@@ -257,29 +309,86 @@ AnalyserEquationAstPtr AnalyserInternalEquation::parseSymEngineExpression(const
...
-        ast->setLeftChild(parseSymEngineExpression(children[0], ast, variableMap));
-        if (children.size() > 1) {
-            ast->setRightChild(parseSymEngineExpression(children[1], ast, variableMap));
+    auto children = seExpression->get_args();
+    auto currentAst = ast;
+
+    // All children except the last are to be assigned as left children in the AST tree.
+    for (int i = 0; i + 1 < children.size(); ++i) {
+        auto childSeExpression = children[i];
+        AnalyserEquationAstPtr childAst = parseSymEngineExpression(childSeExpression, currentAst, variableMap);
```
Comment (agarny) [inline src/analyser.cpp]: Just have `auto childAst = parseSymEngineExpression(children[i], currentAst, variableMap);`

File: src/analyser.cpp [hunk 2]
```diff
@@ -257,29 +309,86 @@ AnalyserEquationAstPtr AnalyserInternalEquation::parseSymEngineExpression(const
...
+            currentAst->setRightChild(newAst);
+            newAst->setParent(currentAst);
+            currentAst = newAst;
         }
     }
 
+    // Set the final node as the right child.
+    if (children.size() >= 2) {
+        auto childSeExpression = children.back();
+        AnalyserEquationAstPtr childAst = parseSymEngineExpression(childSeExpression, currentAst, variableMap);
```
Comment (agarny) [inline src/analyser.cpp]: Just have `auto childAst = parseSymEngineExpression(children.back(), currentAst, variableMap);`

File: tests/resources/analyser/symengine/unarranged_polynomials.cellml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        <apply>
+          <plus/>
+          <apply><power/><ci>b</ci><cn cellml:units="dimensionless">3</cn></apply>
+          <cn cellml:units="dimensionless">4</cn>
+        </apply>
+        <cn cellml:units="dimensionless">12</cn>
+      </apply>
+
+      <!-- Equation 3: c^3 + 3*c = 30 -->
+      <!-- Unique solution: c = 3 -->
```
Comment (agarny) [inline tests/resources/analyser/symengine/unarranged_polynomials.cellml]: This is wrong. If we want `c = 3` then the equation should read `c^3 + 3*c = 36`.

File: tests/resources/analyser/symengine/unarranged_combination.cellml
Comment (agarny) [inline tests/resources/analyser/symengine/unarranged_combination.cellml:1]: How are these test equations relevant? They already solve for `a`, `b`, `c`, and `d`, respectively.

**Review by agarny — CHANGES_REQUESTED** [review]
- In your tests, in addition to confirming that a model is of algebraic type, it would be good to confirm that its equation looks the way we would expect. For this, you could use `Generator::equationCode()`.
- No need to have our CellML files named `unarranged_xxx.cellml`. They can simply be named `xxx.cellml` since they are located under the `symengine` folder.
- For the different equations you are testing, it would be nice to have a comment on why they are being tested.

File: src/analyser.cpp
Comment (RLee64) [reply src/analyser.cpp:202]: I think (almost) all types should be added at some point, it's just that I've been gradually going through them to make sure I'm correctly converting to and from these expressions.

The only one I'm not so sure about are piecewise functions, I highly doubt rearranging for those is supported by symengine (considering how symengine is unable to inverse transcedental functions). It doesn't feel like there's much point for us to spend time trying to rearrange something if it's just going to fail.

I also want to do a bit of testing on differential functions to make sure things function as expected.

File: tests/resources/analyser/symengine/unarranged_combination.cellml
Comment (RLee64) [reply tests/resources/analyser/symengine/unarranged_combination.cellml:1]: I haven't gotten around to cleaning those test cases yet.

As a matter of fact, I'll probably delete all test cases I haven't reviewed yet will just add them back when I actually need them.

Comment (agarny) [reply tests/resources/analyser/symengine/unarranged_combination.cellml:1]: Yes, a PR should NOT contain unfinished / unneeded code.

File: src/analyser.cpp
Comment (agarny) [reply src/analyser.cpp:202]: That's kind of fine but then you should make it clear in the PR. In fact, a PR should be linked to a GitHub issue that describes what we want/need to do and where we can discuss things prior implementation. From where I am standing, your PR has an unclear objective.

File: src/analyser.cpp (hunk 1 above)
Comment (RLee64) [reply src/analyser.cpp]: 3a553bed57b4b83352c4fb2bb144944e0672100b

File: src/analyser.cpp (hunk 2 above)
Comment (RLee64) [reply src/analyser.cpp]: 3a553bed57b4b83352c4fb2bb144944e0672100b

File: tests/resources/analyser/symengine/unarranged_combination.cellml
Comment (RLee64) [reply tests/resources/analyser/symengine/unarranged_combination.cellml:1]: 9efff06fc79148bdb0cab40ef8532fbcf268bab2

File: tests/resources/analyser/symengine/unarranged_polynomials.cellml (hunk 3 above)
Comment (RLee64) [reply tests/resources/analyser/symengine/unarranged_polynomials.cellml]: f8a39f52c927c58f18b0da5dc070f701f1d65f4e

File: src/analyser.cpp
Comment (RLee64) [reply src/analyser.cpp:202]: Tried to make an update to the PR descrption, hopefully it's sufficient now?

**Review by RLee64 — COMMENTED** [review]
> In your tests, in addition to confirming that a model is of algebraic type, it would be good to confirm that its equation looks the way we would expect. For this, you could use Generator::equationCode().

ec0ace2a3a629ea543ca00e2c7c56daf6f6b57d4

> No need to have our CellML files named unarranged_xxx.cellml. They can simply be named xxx.cellml since they are located under the symengine folder.

f8a39f52c927c58f18b0da5dc070f701f1d65f4e

> For the different equations you are testing, it would be nice to have a comment on why they are being tested.

fc1cf1c94c935dfda779e6eb16c2cd50b1ed8824

In addition to these changes I've also done the following
- Cleaned up the trigonometric test case so it's only testing for functionality we know will pass in symengine's current state (i.e. no operations needed to inverse functions). d79d8fc36d48476099d099b6fc5b176602af5865
- Fixed an issue where unary functions weren't being properly converted back (this wasn't picked up since we hadn't encountered symengine outputting unary functions until the proper trigonometric test case) aab7d89379906f057c17c6335b71580c71982d93
- Fixed issue where ast children weren't being assigned their parents (this one I'm just embarrased I hadn't noticed sooner) 9c1e0b8917a00a9789ec44f9dfae7fc87fef7e8b

File: src/analyser.cpp (hunk 1 above)
Comment (agarny) [reply src/analyser.cpp]: I would also remove `childSeExpression` which is only used once and is short enough to be inlined.

File: src/analyser.cpp (hunk 2 above)
Comment (agarny) [reply src/analyser.cpp]: I would also remove `childSeExpression` which is only used once and is short enough to be inlined.

File: src/analyser.cpp
Comment (agarny) [reply src/analyser.cpp:202]: Ok, but remember that, in the future, a GitHub issue will first need to be created and then a PR.

File: src/analyser.cpp (hunk 2 above)
Comment (RLee64) [reply src/analyser.cpp]: f50049e79eeb8ac3cfd23ab71e9fadae927dceee

File: src/analyser.cpp (hunk 1 above)
Comment (RLee64) [reply src/analyser.cpp]: f50049e79eeb8ac3cfd23ab71e9fadae927dceee

**Review by agarny — APPROVED** [review]
Approving, but... at some point, we will need to ensure that ALL tests pass, incl. coverage tests, memory checks, etc.
